### DISCUSSION TITLE
For the id primary key field, should the ID type be used as the identifier instead of using a specific String or Int?

### DISCUSSION THREAD

**Discussion by tianlongJ** [graphql-java/graphql-java#3608] [General]
When using ID as a type identifier, the same code returns different types in different environments. One environment is String, and the other environment is Int.

**Comment by bbakerman** [comment]
Sorry but there is not enough information here for us to help or even consider this an issue.  Can you please expand the question and give examples please.

**Comment by tianlongJ** [comment]
> Sorry but there is not enough information here for us to help or even consider this an issue. Can you please expand the question and give examples please.

This is just a query, not a bug. I just found a type problem when using the ID type.

**Comment by bbakerman** [comment]
I converted this to a discussion if its not a bug.  I still cant say I have enough information to understand the question however.

I dont undersand what you mean by "the same code returns different types in different environments"

Can you please be more descriptive about the code in question, the graphql types used and the scenarios

**Reply by tianlongJ** [reply]
For example, I used the ID type for the id field, which corresponds to the Long type id field in my Java entity. In one of my environments, the return value id type for this request is a string, such as "1", but in another environment, it returns the number 1 without quotes.

**Reply by bbakerman** [reply]
ok I think I understand more

I dont think this is a graphql-java thing

`graphql.scalar.GraphqlIDCoercing#serializeImpl` has this implementation

```
    private String serializeImpl(Object input, @NotNull Locale locale) {
        String result = String.valueOf(input);
        if (result == null) {
            throw new CoercingSerializeException(
                    "Expected type 'ID' but was '" + typeName(input) + "'."
            );
        }
        return result;
    }

```

That is the ID scalars coercing code ALWAYS serialises as a String.  ALWAYS by specification and ALWAYS by implementation

That said its possible your JSON serialisation code might change the value `"1234"` into `1234` say as it goes into the network wire.  This is a major guess but I do recall someone having this problem before

**Reply by tianlongJ** [reply]
That is to say, ID=String, right? As for why such a thing happened, I don't have any good ideas. The environment has been running for many years. This is the first time I have found such a problem. I don't know if it has anything to do with the server/nginx.

**Reply by bbakerman** [reply]
Can you debug it - can you log the graphql java execution result BEFORE it goes into the network layer where it gets converted to JSON say?  Are you using JSON or YAML or what on the wire?

**Reply by tianlongJ** [reply]
No, the problem was discovered in the production environment, and there is no way to debug it
